# Lesson 3 Working with Data

These exercises are to help you check your understanding during class. They will not be collected.

## Useful links
### Class notes
Link on NYU Classes or<br>
Tiny URL: 

### Poll Everywhere
The following links are also on NYU Classes<br>
Login: https://www.polleverywhere.com/auth/saml/nyu_main<br>
Respond: https://pollev.com/clarkson<br>

## Part 1: Lists

### Exercise 0 - Access notebook and copy it
* Login to the JupyterHub: [https://coreua-111001-spring.rcnyu.org/](https://coreua-111001-fall.rcnyu.org/)
* Nacigate to **/shared/class_notebooks** and open lesson03.jpynb
* Save your own copy of this notebook by using the menu opetion File > Save As. In the name field, write lesson03 and then press Save. 
* Now, **refresh the page**. The red icon in the upper right should no longer be visible. 

### Exercise 1 - Review of lists

Last class, we learned to make lists using both c() and number : number. We also learned to do perform arithemtic with lists.

Without using Copy & Paste create a list whose entries are the following numbers:
10 15 20 25 30 35 40 45 50 55.

In [2]:
2:11

[1]  2  3  4  5  6  7  8  9 10 11

In [1]:
2:11*5

[1] 10 15 20 25 30 35 40 45 50 55

### Exercise 2 - Ice cream sale!!!

An icecream truck sells many different types of ice cream. The prices of each type of ice cream are stored in a list called `ice_cream_prices`in the code cell below. Today the icecream is on sale! The amount that each item is discounted is stored in the list called `discounts`. The order of the discounts corresponds to the order of the prices. 

Create a new list with the sale prices of each item and store the list in a variable called `sale_prices`.

In [3]:
ice_cream_prices <- c(1.25, 3, 5.99, 1.75, 6.75, 4.99, 7.09, 3.55, 2)
discounts <- c(0.25, 1, 0.50, 0.5, 1, 0, 1, 0.25, 0)

In [4]:
my_vect # undefined variable

ERROR: Error in eval(expr, envir, enclos): object 'my_vect' not found


In [7]:
# Recall that the pound sign is used for comments

sale_prices <- ice_cream_prices - discounts # This line creates the list and stores it.
sale_prices # This line is what causes R to output the list.

[1] 1.00 2.00 5.49 1.25 5.75 4.99 6.09 3.30 2.00

### Exercises 3- More ice cream 
Use `sale_prices` from excercise 2 and R functions to determine the ansers to the following questions:
* How many types of ice cream does the truck sell?
* What is the cost of the most exspensive frozen treat?
* What is the cost of the cheapest ice cream option?
* What is the average cost of an item from the truck rounded to the nearest cent?

In [11]:
length(sale_prices)

[1] 9

There are 9 types of ice cream.

In [17]:
max(sale_prices)

[1] 6.09

The cost of the most exspensive treat is $6.09.

In [18]:
min(sale_prices)

[1] 1

The cost of the least exspensive treat is $1.

In [9]:
mean(sale_prices) # average price

[1] 3.541111

In [19]:
round(mean(sale_prices),2) # two decimal places

[1] 3.54

The average cost of ice cream is $3.54.

## Part 2: Old faithful

### Exercise 4 - Old Failthful

One of the data sets in the datasets package is called `faithful`. Use the documentation on this dataset to answer the following question. Wrtie your answer in the markdown cell below. 

Why is there reason to believe that this data includes errors introduced by humans?

In [15]:
?faithful # question mark for documentation

Eruption times are very often multiples of five. 

### Exercise 5 - The last observation

How long was the last erruption recorded in the fialthful data set? Write you answer in the markdown cell below. Don't forget units!

In [20]:
faithful # display the whole data frame and scoll to the bottom

eruptions waiting
1   3.600     79     
2   1.800     54     
3   3.333     74     
4   2.283     62     
5   4.533     85     
6   2.883     55     
7   4.700     88     
8   3.600     85     
9   1.950     51     
10  4.350     85     
11  1.833     54     
12  3.917     84     
13  4.200     78     
14  1.750     47     
15  4.700     83     
16  2.167     52     
17  1.750     62     
18  4.800     84     
19  1.600     52     
20  4.250     79     
21  1.800     51     
22  1.750     47     
23  3.450     78     
24  3.067     69     
25  4.533     74     
26  3.600     83     
27  1.967     55     
28  4.083     76     
29  3.850     78     
30  4.433     79     
... ...       ...    
243 4.933     86     
244 2.900     63     
245 4.583     85     
246 3.833     82     
247 2.083     57     
248 4.367     82     
249 2.133     67     
250 4.350     74     
251 2.200     54     
252 4.450     83     
253 3.567     73     
254 4.500     73     
255 4.150     88     
256 3.817     80     
257 3.917     71     
258 4.450     83     
259 2.000     56     
260 4.283     79     
261 4.767     78     
262 4.533     84     
263 1.850     58     
264 4.250     83     
265 1.983     43     
266 2.250     60     
267 4.750     75     
268 4.117     81     
269 2.150     46     
270 4.417     90     
271 1.817     46     
272 4.467     74

The last recorded eruption was 4.467 minutes long.

In [21]:
dim(faithful)

[1] 272   2

In [16]:
faithful[272,1] # row, column

[1] 4.467

### Exercise 6 - Change the units

The faithful data set gives the duration of the erruptions in minutes and fractions of a minute. It would be easer to understad the data if we converted the decimals to seconds. Demonstrate that this could be done by finding the duration of the 13th recorded eruption in therms of minuts and seconds. 

In [1]:
faithful[13,1] # 13th row first column

[1] 4.2

In [2]:
# We see that the duration is 4.2 minutes. We need to convert 0.2 minutes into seconds
0.2*60

[1] 12

The duration is 4 minutes and 12 seconds.

## Part 3: UC Berkeley

### Exercise 7 - Load Berkeley addmissions data

Execute the command in the code cell below to load the Berkeley admisions data. Then varify that it loaded correctly by excuting a command that displays all or part of the data. 

In [5]:
berkeley <- read.csv('~/shared/data/berkeley73.csv')

In [6]:
berkeley <- read.csv('~/shared/data/berkeley73.csv')
berkeley

Department Men_Applicants Men_Admitted Women_Applicants Women_Admitted
1 A          825            512          108               89           
2 B          560            353           25               17           
3 C          325            120          593              202           
4 D          417            138          375              131           
5 E          191             53          393               94           
6 F          373             22          341               24

In [14]:
# Strategy 1
berkeley

Department Men_Applicants Men_Admitted Women_Applicants Women_Admitted
1 A          825            512          108               89           
2 B          560            353           25               17           
3 C          325            120          593              202           
4 D          417            138          375              131           
5 E          191             53          393               94           
6 F          373             22          341               24

In [13]:
# Strategy 2
head(berkeley,2)

Department Men_Applicants Men_Admitted Women_Applicants Women_Admitted
1 A          825            512          108              89            
2 B          560            353           25              17

### Exercise 8 - Admissons rates

Add two new columns to the berkeley data frame. One that contains the percentage of men who were admitted by each department and a second that contains the percentage of women that were admitted by each department. 

In [18]:
# percent admitted = total admiteed/ total applied * 100
berkeley$Per_Men_Admitted <- berkeley$Men_Admitted/berkeley$Men_Applicants*100
berkeley$Per_Women_Admitted <- berkeley$Women_Admitted/berkeley$Women_Applicants*100
berkeley

Department Men_Applicants Men_Admitted Women_Applicants Women_Admitted
1 A          825            512          108               89           
2 B          560            353           25               17           
3 C          325            120          593              202           
4 D          417            138          375              131           
5 E          191             53          393               94           
6 F          373             22          341               24           
  Per_Men_Admitted Per_Women_Admitted
1 62.060606        82.407407         
2 63.035714        68.000000         
3 36.923077        34.064081         
4 33.093525        34.933333         
5 27.748691        23.918575         
6  5.898123         7.038123

## Examples

In [1]:
mean(3:10)

[1] 6.5

### Example 1 - Variables and list fucntions

In [33]:
numbers <- c(5.8, 3.52, 7.81, 0.4, 4.79, 3.04, 4.89, 5.23, 2.37, 4.32) # generated using a random number generator
mean(numbers)

[1] 4.217

In [35]:
max(numbers)

[1] 7.81

### Example 2 - Airquality

In [8]:
library(help = "datasets")

In [17]:
?airquality # gives the documentation for this BUILT-IN dataset

In [13]:
head(airquality,3)

Ozone Solar.R Wind Temp Month Day
1 41    190      7.4 67   5     1  
2 36    118      8.0 72   5     2  
3 12    149     12.6 74   5     3

In [3]:
str(airquality) # structure function

'data.frame':	153 obs. of  6 variables:
 $ Ozone  : int  41 36 12 18 NA 28 23 19 8 NA ...
 $ Solar.R: int  190 118 149 313 NA NA 299 99 19 194 ...
 $ Wind   : num  7.4 8 12.6 11.5 14.3 14.9 8.6 13.8 20.1 8.6 ...
 $ Temp   : int  67 72 74 62 56 66 65 59 61 69 ...
 $ Month  : int  5 5 5 5 5 5 5 5 5 5 ...
 $ Day    : int  1 2 3 4 5 6 7 8 9 10 ...


In [14]:
mean(airquality$temp) # Note that R is sensitice to campialization

Warning message in mean.default(airquality$temp):
“argument is not numeric or logical: returning NA”

[1] NA

In [15]:
mean(airquality$Temp)

[1] 77.88235

The average high temperature in NYC from May to Septemerin 1973 was $77.9^\circ$ F.

In [18]:
max(airquality$Temp)

[1] 97

The hottest day was $97^\circ$ F.

### Example 3 - Change the units

We can use the formula below to convert a temperature $f$ in degrees Fahrenheit to the equivalent temperature $c$ in degrees Celcius
$$(f − 32) \cdot \frac{5}{9} = c$$

In [19]:
f <- 97
(f-32)*5/9

[1] 36.11111

The hottest day was $36.1^\circ$ C.

In [23]:
airquality$Celcius <- (airquality$Temp - 32)*5/9
head(airquality)

Ozone Solar.R Wind Temp Month Day Celcius 
1 41    190      7.4 67   5     1   19.44444
2 36    118      8.0 72   5     2   22.22222
3 12    149     12.6 74   5     3   23.33333
4 18    313     11.5 62   5     4   16.66667
5 NA     NA     14.3 56   5     5   13.33333
6 28     NA     14.9 66   5     6   18.88889